librocksdb_sys/crocksdb/c.cc

```diff
@@ -3779,4 +3798,108 @@ int crocksdb_keyversions_type(const crocksdb_keyversions_t *kvs, int index) {
   return kvs->rep[index].type;
 }
 
+struct ExternalSstFileModifier {
+  ExternalSstFileModifier(Env *env, ColumnFamilyData *cfd)
+  :env_(env), cfd_(cfd), table_reader_(nullptr) { }
+
+  Status Open(std::string file) {
+    file_ = file;
+    // Get External Sst File Size
+    uint64_t file_size;
+    auto status = env_->GetFileSize(file_, &file_size);
+    if (!status.ok()) {
+      return status;
+    }
+
+    // Open External Sst File
+    std::unique_ptr<RandomAccessFile> sst_file;
+    std::unique_ptr<RandomAccessFileReader> sst_file_reader;
+    status = env_->NewRandomAccessFile(file_, &sst_file, env_options_);
+    if (!status.ok()) {
+      return status;
+    }
+    sst_file_reader.reset(new RandomAccessFileReader(std::move(sst_file), file_));
+
+    // Get Table Reader
+    status = cfd_->ioptions()->table_factory->NewTableReader(
+        TableReaderOptions(*cfd_->ioptions(), env_options_,
+                           cfd_->internal_comparator()),
+        std::move(sst_file_reader), file_size, &table_reader_);
+    return status;
+  }
+
+  Status ModifySeqNo(uint64_t seq_no) {
+    if (table_reader_ == nullptr) {
+      return Status::InvalidArgument("File is not open or seq-no has been modified");
+    }
+    // Get the external file properties
+    auto props = table_reader_->GetTableProperties();
+    const auto& uprops = props->user_collected_properties;
+    // Validate version and seqno offset
+    auto version_iter = uprops.find(ExternalSstFilePropertyNames::kVersion);
+    if (version_iter == uprops.end()) {
+      return Status::Corruption("External file version not found");
+    }
+    uint32_t version = DecodeFixed32(version_iter->second.c_str());
+    if (version != 2) {
+      return Status::NotSupported("External file version should be 2");
+    }
+
+    auto seqno_iter = uprops.find(ExternalSstFilePropertyNames::kGlobalSeqno);
+    uint64_t ori_seq_no = DecodeFixed64(seqno_iter->second.c_str());
+    uint64_t offset = props->properties_offsets.at(ExternalSstFilePropertyNames::kGlobalSeqno);
+    if (offset == 0) {
+      return Status::Corruption("Was not able to find file global seqno field");
+    }
+
+    if (ori_seq_no == seq_no) {
+      // This file already have the correct global seqno
+      return Status::OK();
+    }
+
+    std::unique_ptr<RandomRWFile> rwfile;
+    auto status = env_->NewRandomRWFile(file_, &rwfile, env_options_);
+    if (!status.ok()) {
+      return status;
+    }
+
+    // Write the new seqno in the global sequence number field in the file
+    std::string seqno_val;
+    PutFixed64(&seqno_val, seq_no);
+    status = rwfile->Write(offset, seqno_val);
+    if (status.ok()) {
+      table_reader_ = nullptr;
+    }
+    return status;
+  }
+
+  private:
+    EnvOptions        env_options_;
```

Comment:
`env_options_ ` is related to `env_`, also need to be initialed relatively.

Reply:
`EnvOptions` is initialized with `DBOptions` only, and I can't use DBOptions to initialize it because this will cause compilation error.
In `SstDumpTool`, `EnvOptions` also uses default settings, and the only difference is that `SstDumpTool` sets `env_options_.use_mmap_reads =  true` when the table is `PlainTable`, since we do not use `PlainTable`, so to make things simple I made this `modifier` only support block based table, otherwise I have to introduce several other non-public API.

Reply:
`EnvOptions` can only be initialized by `DBOptions`.